contracts/integrations/contracts/dydx/ZeroExV3ExchangeWrapper.sol

```diff
@@ -0,0 +1,329 @@
+import { IExchange } from "@0x/contracts-exchange/contracts/src/interfaces/IExchange.sol";
+import { LibFillResults } from "@0x/contracts-exchange-libs/contracts/src/LibFillResults.sol";
+import { LibOrder } from "@0x/contracts-exchange-libs/contracts/src/LibOrder.sol";
+import { ExchangeReader } from "./interfaces/ExchangeReader.sol";
+import { ExchangeWrapper } from "./interfaces/ExchangeWrapper.sol";
+import { AdvancedTokenInteract } from "./libs/AdvancedTokenInteract.sol";
+import { MathHelpers } from "./libs/MathHelpers.sol";
+import { TokenInteract } from "./libs/TokenInteract.sol";
+
+
+/**
+ * @title ZeroExV3ExchangeWrapper
+ * @author dYdX/0x
+ *
+ * dYdX ExchangeWrapper to interface with 0x Version 3
+ */
+contract ZeroExV3ExchangeWrapper is
+    ExchangeWrapper,
+    ExchangeReader
+{
+    using SafeMath for uint256;
+    using TokenInteract for address;
+    using AdvancedTokenInteract for address;
+
+    // ============ Constants ============
+
+    // maximum gas price, 30 Gwei (choose value for deployment)
+    uint256 constant MAX_GAS_PRICE = 30000000000 wei;
+
+    // ============ State Variables ============
+
+    // msg.senders that will put the correct tradeOriginator in callerData when doing an exchange
+    mapping (address => bool) public TRUSTED_MSG_SENDER;
+
+    // address of the ZeroEx V3 Exchange
+    address public ZERO_EX_EXCHANGE;
+
+    // address of the ZeroEx ERC20Proxy
+    address public ZERO_EX_TOKEN_PROXY;
+
+    // address of the WETH token contract
+    address public WETH_TOKEN;
+
+    // ============ Constructor ============
+
+    constructor(
+        address zeroExExchange,
+        address zeroExProxy,
+        address weth,
+        address[] memory trustedMsgSenders
+    )
+        public
+    {
+        ZERO_EX_EXCHANGE = zeroExExchange;
+        ZERO_EX_TOKEN_PROXY = zeroExProxy;
+        WETH_TOKEN = weth;
+
+        for (uint256 i = 0; i < trustedMsgSenders.length; i++) {
+            TRUSTED_MSG_SENDER[trustedMsgSenders[i]] = true;
+        }
+    }
+
+    // ============ Public Functions ============
+
+    function exchange(
+        address tradeOriginator,
+        address receiver,
+        address makerToken,
+        address takerToken,
+        uint256 requestedFillAmount,
+        bytes calldata orderData
+    )
+        external
+        returns (uint256)
+    {
+        // prepare the exchange
+        (LibOrder.Order memory order, address takerFeeToken) = parseOrder(orderData, makerToken, takerToken);
+        bytes memory signature = parseSignature(orderData);
+
+        // transfer taker fee from trader if applicable
+        transferTakerFee(
+            order,
+            tradeOriginator,
+            requestedFillAmount,
+            takerFeeToken
+        );
+
+        // set protocol fee allowance
+        IExchange v3Exchange = IExchange(ZERO_EX_EXCHANGE);
+        address protocolFeeCollector = v3Exchange.protocolFeeCollector();
```

Comment:
@jalextowle suggested exposing an external function that would allow anyone (us, probably) to do this (call `v3Exchange.protocolFeeCollector()`, set the allowance). How do you feel about that approach?